# Bike Sharing Demand Forecasting and Business Insights Using Linear Regression

## Business Problem

A bike-sharing company wants to better understand customer demand to optimize bike availability, staffing, and operational planning.

The objective of this project is to identify the factors that influence bike rental demand and build a predictive model that can forecast future rentals.

This analysis will help decision-makers improve resource allocation and customer satisfaction.

# Project Workflow

1. Data Understanding

2. Data Cleaning

3. Exploratory Data Analysis

4. Feature Engineering

5. Linear Regression Modeling

6. Model Evaluation

7. Business Recommendations

In [2]:
import pandas as pd

In [6]:
df = pd.read_csv("day.csv")

In [3]:
import statsmodels.formula.api as smf

In [7]:
print("First 5 rows:")
display(df.head())

First 5 rows:


,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [8]:
print("\nDataset Information:")
df.info()



Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


In [9]:
print("\nSummary Statistics:")
display(df.describe())


Summary Statistics:


,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


In [10]:
df['dteday'] = pd.to_datetime(df['dteday'])

In [11]:
df['month_name'] = df['dteday'].dt.month_name()

In [12]:
df[['dteday', 'month_name']].head()

,dteday,month_name
0,2011-01-01,January
1,2011-01-02,January
2,2011-01-03,January
3,2011-01-04,January
4,2011-01-05,January


In [13]:
Model1 = smf.ols('cnt ~ month_name', data=df).fit()



In [14]:
print(Model1.summary())

                            OLS Regression Results                            
Dep. Variable:                    cnt   R-squared:                       0.391
Model:                            OLS   Adj. R-squared:                  0.381
Method:                 Least Squares   F-statistic:                     41.90
Date:                Fri, 12 Jun 2026   Prob (F-statistic):           4.25e-70
Time:                        18:40:52   Log-Likelihood:                -6388.6
No. Observations:                 731   AIC:                         1.280e+04
Df Residuals:                     719   BIC:                         1.286e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                4

R_squared is 0.391 means that the month explains 40% of the variation.

April is the reference month

The reference month' predicted cnt is almost 4485

Predicted cnt for January 4485 + 1287 = 5772


Predicted cnt for June    4485 + (-2309) = 2176

In [15]:
Model2 = smf.ols('cnt ~ temp + month_name', data=df).fit()

In [16]:
print(Model2.summary())

                            OLS Regression Results                            
Dep. Variable:                    cnt   R-squared:                       0.447
Model:                            OLS   Adj. R-squared:                  0.438
Method:                 Least Squares   F-statistic:                     48.35
Date:                Fri, 12 Jun 2026   Prob (F-statistic):           4.26e-84
Time:                        18:48:37   Log-Likelihood:                -6353.2
No. Observations:                 731   AIC:                         1.273e+04
Df Residuals:                     718   BIC:                         1.279e+04
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                1

The R-squared value for Model2 is 0.447


Adding the predictor temp to the model provides more information, and explains more the variation in cnt, since the temperature is an important factor that affects the bike rentals.

The coefficient estimate for January model1 is -2309


The coefficient estimate for January model2 is -852


Model2 includes both month_name and temperature (temp) as predictors. Since temperature is strongly related to bike rentals and January is typically a colder month, some of the variation previously attributed to January is now explained by temperature. As a result, the January coefficient becomes smaller in magnitude, changing from -2309 to -852.

This indicates that part of the lower bike rental count observed in January is due to colder temperatures rather than the month itself.

Following module 2,

Predicted cnt = 1455 +(6235 × 0.25)− 852 = 2261

# Key Insights

1. Temperature is a major driver of bike rental demand. Adding temperature to the model improved its explanatory power.

2. The final model explains approximately 44.7% of the variation in bike rental demand (R² = 0.447).

3. January initially appeared to strongly reduce bike demand, but much of this effect was explained by colder temperatures.

4. Seasonal and weather-related factors play an important role in customer behavior.

5. Combining multiple predictors provides better demand forecasts than using a single variable alone.

# Business Recommendations

- Increase bike availability during warmer periods when demand is expected to rise.

- Use weather forecasts to proactively adjust bike distribution across stations.

- Schedule maintenance activities during colder months when demand is lower.

- Use predictive models to support staffing and operational planning.

- Continue improving the model by incorporating additional variables such as humidity, wind speed, holidays, and working days.

# Project Summary

This project demonstrates how machine learning can be used to forecast bike-sharing demand and support business decision-making.

By analyzing seasonal and weather-related factors, we identified key drivers of customer behavior and built regression models to predict future demand.

The results can help a bike-sharing company improve resource allocation, staffing decisions, and operational efficiency.